File: src/crossalpha/observatory/live_health.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from crossalpha.domain.models import RawSnapshotManifest
from crossalpha.observatory.health import DEFAULT_EXPECTED_SERIES, observatory_health, verify_snapshot


def _state_path(data_root: Path, source_id: str, observation_type: str) -> Path:
    source = source_id.replace("/", "_").replace("..", "_")
    observation = observation_type.replace("/", "_").replace("..", "_")
    return data_root / "manifests" / "series" / source / f"{observation}.json"


def observatory_live_health(
    data_root: Path,
    *,
    stale_after_seconds: int = 900,
    expected_series: tuple[tuple[str, str], ...] = DEFAULT_EXPECTED_SERIES,
    verify_latest: bool = True,
    now: datetime | None = None,
) -> dict[str, Any]:
    """O(1)-per-series health check using rebuildable incremental series state.

    Falls back to the full audit-ledger scan until O0.2 indexes have been built.
    Historical gap analysis deliberately stays in `observatory_health`; this path is
    for watchdog freshness/integrity checks and remains fast as the audit ledger grows.
    """
    now = now or datetime.now(timezone.utc)
    state_paths = [_state_path(data_root, source, observation) for source, observation in expected_series]
    if not all(path.exists() for path in state_paths):
        report = observatory_health(
            data_root,
            stale_after_seconds=stale_after_seconds,
            expected_series=expected_series,
            verify_latest=verify_latest,
            now=now,
        )
        report["mode"] = "full_manifest_fallback"
        return report

    current_ok = True
    series_report: dict[str, Any] = {}
    total_records = 0
    total_raw_bytes = 0
    compression_sample_records = 0
    compression_sample_raw_bytes = 0
    total_compressed_bytes = 0

    for (source_id, observation_type), state_path in zip(expected_series, state_paths, strict=True):
        state = json.loads(state_path.read_text(encoding="utf-8"))
        latest_manifest = RawSnapshotManifest.model_validate(state["latest_manifest"])
        latest_observed = latest_manifest.observed_at
        if latest_observed.tzinfo is None:
            latest_observed = latest_observed.replace(tzinfo=timezone.utc)
        age_seconds = max((now - latest_observed).total_seconds(), 0.0)
        stale = age_seconds > stale_after_seconds
        integrity = verify_snapshot(latest_manifest) if verify_latest else None
        series_ok = not stale and (integrity is None or integrity["ok"])
        current_ok = current_ok and series_ok

        count = int(state.get("count", 0))
        raw_bytes = int(state.get("raw_bytes_total", 0))
        sample_records = int(state.get("compression_sample_records", 0))
        sample_raw = int(state.get("compression_sample_raw_bytes", 0))
        compressed = int(state.get("compressed_bytes_total", 0))
        total_records += count
        total_raw_bytes += raw_bytes
        compression_sample_records += sample_records
        compression_sample_raw_bytes += sample_raw
        total_compressed_bytes += compressed

        label = f"{source_id}/{observation_type}"
        series_report[label] = {
            "ok": series_ok,
            "count": count,
            "latest_observed_at": latest_observed.isoformat(),
            "age_seconds": round(age_seconds, 3),
            "stale": stale,
            "last_interval_seconds": state.get("last_interval_seconds"),
            "max_interval_seconds": state.get("max_interval_seconds"),
            "latest_integrity": integrity,
        }

    compression_ratio = None
    if compression_sample_raw_bytes and total_compressed_bytes:
        compression_ratio = compression_sample_raw_bytes / total_compressed_bytes

    return {
        "ok": current_ok,
        "mode": "series_state",
        "checked_at": now.isoformat(),
        "manifest_records": total_records,
        "manifest_errors": [],
        "stale_after_seconds": stale_after_seconds,
        "raw_bytes_manifested": total_raw_bytes,
        "compression_sample_records": compression_sample_records,
        "compression_sample_raw_bytes": compression_sample_raw_bytes,
        "compressed_bytes_manifested": total_compressed_bytes,
        "compression_ratio": round(compression_ratio, 3) if compression_ratio else None,
        "series": series_report,
    }

File: crates/crossalpha-observatory/src/live_health.rs
use anyhow::Context;
use anyhow::Result;
use chrono::{DateTime, SecondsFormat, Timelike, Utc};
use crossalpha_storage::SeriesState;
use serde_json::Map;
use serde_json::Value;
use serde_json::json;
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use crate::health::DEFAULT_EXPECTED_SERIES;
use crate::health::observatory_health;
use crate::health::verify_snapshot;

pub fn observatory_live_health(
    data_root: &Path,
    stale_after_seconds: u64,
    verify_latest: bool,
    now: Option<DateTime<Utc>>,
) -> Result<Value> {
    let now = now.unwrap_or_else(Utc::now);
    let state_paths: Vec<PathBuf> = DEFAULT_EXPECTED_SERIES
        .iter()
        .map(|(source, observation)| state_path(data_root, source, observation))
        .collect();

    if !state_paths.iter().all(|path| path.exists()) {
        let report = observatory_health(
            data_root,
            300,
            stale_after_seconds,
            verify_latest,
            Some(now),
        )?;
        let mut value = serde_json::to_value(report)?;
        let object = value
            .as_object_mut()
            .context("full health report is not a JSON object")?;
        object.insert(
            "mode".to_owned(),
            Value::String("full_manifest_fallback".to_owned()),
        );
        return Ok(value);
    }

    let mut current_ok = true;
    let mut series_report = BTreeMap::new();
    let mut total_records = 0u64;
    let mut total_raw_bytes = 0u64;
    let mut compression_sample_records = 0u64;
    let mut compression_sample_raw_bytes = 0u64;
    let mut total_compressed_bytes = 0u64;

    for ((source_id, observation_type), path) in
        DEFAULT_EXPECTED_SERIES.iter().zip(state_paths.iter())
    {
        let state: SeriesState = serde_json::from_slice(&fs::read(path)?)
            .with_context(|| format!("parse series state {}", path.display()))?;
        let latest_manifest = state
            .latest_manifest
            .as_ref()
            .with_context(|| format!("series state missing latest_manifest: {}", path.display()))?;
        let latest_observed = latest_manifest.observed_at;
        let age_seconds = duration_seconds(now - latest_observed).max(0.0);
        let stale = age_seconds > stale_after_seconds as f64;
        let integrity = verify_latest.then(|| verify_snapshot(latest_manifest));
        let integrity_ok = integrity
            .as_ref()
            .and_then(|value| value.get("ok"))
            .and_then(Value::as_bool)
            .unwrap_or(true);
        let series_ok = !stale && integrity_ok;
        current_ok = current_ok && series_ok;

        total_records += state.count;
        total_raw_bytes += state.raw_bytes_total;
        compression_sample_records += state.compression_sample_records;
        compression_sample_raw_bytes += state.compression_sample_raw_bytes;
        total_compressed_bytes += state.compressed_bytes_total;

        let label = format!("{source_id}/{observation_type}");
        series_report.insert(
            label,
            json!({
                "ok": series_ok,
                "count": state.count,
                "latest_observed_at": python_isoformat(latest_observed),
                "age_seconds": round3(age_seconds),
                "stale": stale,
                "last_interval_seconds": state.last_interval_seconds,
                "max_interval_seconds": state.max_interval_seconds,
                "latest_integrity": integrity,
            }),
        );
    }

    let compression_ratio = if compression_sample_raw_bytes > 0 && total_compressed_bytes > 0 {
        Some(round3(
            compression_sample_raw_bytes as f64 / total_compressed_bytes as f64,
        ))
    } else {
        None
    };

    let mut root = Map::new();
    root.insert("ok".to_owned(), Value::Bool(current_ok));
    root.insert("mode".to_owned(), Value::String("series_state".to_owned()));
    root.insert(
        "checked_at".to_owned(),
        Value::String(python_isoformat(now)),
    );
    root.insert("manifest_records".to_owned(), json!(total_records));
    root.insert("manifest_errors".to_owned(), json!([]));
    root.insert("stale_after_seconds".to_owned(), json!(stale_after_seconds));
    root.insert("raw_bytes_manifested".to_owned(), json!(total_raw_bytes));
    root.insert(
        "compression_sample_records".to_owned(),
        json!(compression_sample_records),
    );
    root.insert(
        "compression_sample_raw_bytes".to_owned(),
        json!(compression_sample_raw_bytes),
    );
    root.insert(
        "compressed_bytes_manifested".to_owned(),
        json!(total_compressed_bytes),
    );
    root.insert("compression_ratio".to_owned(), json!(compression_ratio));
    root.insert("series".to_owned(), serde_json::to_value(series_report)?);
    Ok(Value::Object(root))
}

fn state_path(data_root: &Path, source_id: &str, observation_type: &str) -> PathBuf {
    let source = safe_component(source_id);
    let observation = safe_component(observation_type);
    data_root
        .join("manifests")
        .join("series")
        .join(source)
        .join(format!("{observation}.json"))
}

fn safe_component(value: &str) -> String {
    value.replace('/', "_").replace("..", "_")
}

fn duration_seconds(duration: chrono::Duration) -> f64 {
    duration
        .num_microseconds()
        .map(|value| value as f64 / 1_000_000.0)
        .unwrap_or_else(|| duration.num_seconds() as f64)
}

fn python_isoformat(timestamp: DateTime<Utc>) -> String {
    let format = if timestamp.nanosecond() == 0 {
        SecondsFormat::Secs
    } else {
        SecondsFormat::Micros
    };
    timestamp.to_rfc3339_opts(format, false)
}

fn round3(value: f64) -> f64 {
    (value * 1000.0).round() / 1000.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn state_path_matches_python_contract() {
        assert_eq!(
            state_path(Path::new("/tmp/data"), "a/b", "x..y"),
            PathBuf::from("/tmp/data/manifests/series/a_b/x_y.json")
        );
    }
}
